articles/tutorial-qdk-grovers-search.md

```diff
@@ -11,49 +11,367 @@ title: Run Grover's search algorithm in Q# - Quantum Development Kit
+1. **Implement the function of the Grover's task as a quantum oracle:** to implement Grover's algorithm, you need to implement the function $f(x)$ of your Grover's task as a [quantum oracle](xref:microsoft.quantum.concepts.oracles).
+1. **Use Grover's algorithm with your oracle to solve the task:** once you have quantum oracle, you can plug it into your Grover's algorithm implementation to solve the problem and interpret the output.
 
-At the heart of the algorithm are two steps that incrementally boost the coefficient of the index that we are looking for, until the probability amplitude of that coefficient approaches one.
+## Quick overview of Grover's algorithm
 
-The number of incremental boosts is fewer than the number of items in the list. This is why Grover's search algorithm performs the search in fewer steps than any classical algorithm.
+Suppose we have $N=2^n$ eligible items for the search task and we index them by assigning each item a integer from $0$ to
+$N-1$. The steps of the algorithm are:
 
-![Functional diagram of Grover's search algorithm](~/media/grover.png)
+1. Start with a register of $n$ qubits initialized in the state $\ket{0}$ by applying $H$ to each qubit of the register.
+1. Prepare the register into a uniform superposition: $$|\psi\rangle=\frac{1}{N^{1 / 2}} \sum_{x=0}^{N-1}|x\rangle$$
+1. Apply $N_{\text{optimal}}$ times the following operations to the register:
+   1. The phase oracle $O_f$ that applies a conditional phase shift of $-1$ for the solution items.
+   1. Apply $H$ to each qubit of the register.
+   1. A conditional phase shift of $-1$ to every computational basis state except $\ket{0}$.
+   1. Apply $H$ to each qubit of the register.
+1. Measure the register to obtain the index of a item that's a solution with very high probability.
+1. Check if it's a valid solution. If not, start again.
 
-## Write the code
+## Write the code for Grover's algorithm
 
-1. Using the Quantum Development Kit, [create a new Q# project for the application](xref:microsoft.quantum.install-qdk.overview.standalone). Title the project `Grover`.
+Now let's see how to implement the code in Q#.
 
-1. Add the following code to the `Program.qs` file in your new project:
+### Grover's diffusion operator
+
+First, we are going to write an operation that applies the steps **b**, **c** and **d** of the loop. These steps
+are sometimes known as the Grover's diffusion operation.
+
+```qsharp
+operation ReflectAboutUniform(inputQubits : Qubit[]) : Unit {
+
+    within {
+        ApplyToEachA(H, inputQubits);
+        ApplyToEachA(X, inputQubits);
+    } apply {
+        Controlled Z(Most(inputQubits), Tail(inputQubits));
+    }
+
+}
+```
 
-    :::code language="qsharp" source="~/quantum/samples/algorithms/simple-grover/SimpleGrover.qs" range="4-":::
```

Comment:
Directly embedding code means that unit tests and CI will no longer ensure that code compiles correctly against the latest version of Q#; if at all possible, I would strongly suggest applying the proposed code changes here to the samples repo so that we can maintain a single and tested source of truth for up-to-date code.

Reply:
This wasn't intended to be a substitution of the samples, but it could be... Maybe after is approved we can substitute the sample or create a new one.

But we had recently a discussion about embedding code (https://github.com/MicrosoftDocs/quantum-docs/issues/28) and we decided to remove the code snippets to avoid incorrect code in the documentation, even if it produces two different sources of content that may get out of sync. Maybe you want to reopen the discussion there.